### ISSUE TITLE
OC: OEM specified values for system-id and SystemSerialNumber are not written to data hub

### ISSUE THREAD

**Issue by Bmju** [acidanthera/bugtracker#1469]
Somewhat related to #1468, I am finding that if I override data hub protocol, specify `PlatformInfo` `Automatic` = `true` with  `UpdateDataHub` = `true`, and specify empty string values in `Generic/SystemUUID` and `Generic/SystemSerialNumber` the no values at all are written to the data hub for `system-id` and `SystemSerialNumber`. I would have expected the OEM specified values to appear. The OEM specified values still work fine in SMBIOS when this is happening. Specifying explicit values works fine for both SMBIOS and data hub.

IORegistryExplorer screenshots and log OCDH lines below.

---

_With explicit values:_

<img width="589" alt="Screenshot 2021-02-03 at 10 26 43" src="https://user-images.githubusercontent.com/25643187/106738094-70539300-660f-11eb-990f-449600eb49ec.png">

```
04:084 00:020 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:name (9) - Success
04:104 00:020 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:Model (30) - Success
04:125 00:020 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:SystemSerialNumber (26) - Success
04:144 00:019 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:system-id (16) - Success
04:163 00:019 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:board-id (21) - Success
04:182 00:019 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:StartupPowerEvents (8) - Success
04:202 00:019 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:InitialTSC (8) - Success
04:221 00:019 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:FSBFrequency (8) - Success
04:240 00:019 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:DevicePathsSupported (4) - Success
04:259 00:019 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:REV (6) - Success
04:278 00:019 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:RBr (8) - Success
04:299 00:020 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:RPlt (8) - Success
```

_With empty string values:_

<img width="589" alt="Screenshot 2021-02-03 at 10 32 39" src="https://user-images.githubusercontent.com/25643187/106738064-6336a400-660f-11eb-8add-79e60591ea92.png">

```
04:062 00:020 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:name (9) - Success
04:082 00:019 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:Model (30) - Success
04:102 00:019 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:board-id (21) - Success
04:121 00:019 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:StartupPowerEvents (8) - Success
04:139 00:018 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:InitialTSC (8) - Success
04:158 00:018 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:FSBFrequency (8) - Success
04:177 00:019 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:DevicePathsSupported (4) - Success
04:196 00:018 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:REV (6) - Success
04:215 00:018 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:RBr (8) - Success
04:234 00:019 OCDH: Setting DataHub 64517CC8-6561-4051-B03C-5964B60F4C7A:RPlt (8) - Success
```

**Comment by vit9696** [comment]
This is expected and defined behaviour. Not setting `SystemUUID` does not mean "Set OEM value". It means _leave intact_. For this reason, once DataHub protocol is replaced, and your system-id value is lost, nothing will set it.

**Comment by Bmju** [comment]
UpdateDataHub
Description: Update Data Hub fields. These fields are read from Generic or DataHub sections depending on Automatic value.

UpdateSMBIOS
Description: Update SMBIOS fields. These fields are read from Generic or SMBIOS sections depending on Automatic value.

SystemSerialNumber
Failsafe: OEM specified or not installed

SystemUUID
Failsafe: OEM specified or not installed

I agree that "OEM specified or not installed" is stated, but I assumed this must mean "OEM specified value will be used if there is one, and the value will not be installed otherwise"?

So, I am intending to correctly use the failsafe values (setting present, empty string) to get the OEM value (if there is one, which there is), and I am intending to use the specified flags to populate the SMBIOS and DataHub with these values.

Put another way, there is certainly nothing to indicate that when using `Automatic` to populate from the `Generic` section, that, for values which exist in both, the values which go into DataHub may be different from the values that go into SMBIOS (more specifically, present in one but not present in the other), when you are asking OC to populate both from the same data.

In addition to the above considerations, for the specific use-case, the ability to read the OEM specified values in the Generic section is currently not also useable in combination with `UpdateDataHub` since these values propagate to SMBIOS but do not propagate to Data Hub. Or, on a positive note, this would be both usable and useful if they _did_ propagate there as well.

**Comment by Bmju** [comment]
@vit9696 Thank you for your patience. You haven't re-opened, so trying to summarise as quickly as possible, I still believe this is wrong.

For clarity, I am talking about the SystemUUID and SystemSerialNumber settings in the Generic section, not in the DataHub section, and about the case when Automatic and UpdateDataHub are both set to true.

- I believe the documentation I quoted above really makes it look like the settings would work as I am expecting
- As an end user, I cannot see any logic to the fact that Generic/SystemUUID is propagated to both SMBIOS and Data Hub _when a value is present_, but only to SMBIOS when empty string is present (empty string is specified by the docs as meaning "Use the OEM specified value if there is one" - and there is one). As a user, I would want the value specified by Generic/SystemUUID (and SystemSerialNumber) to propagate in the same way to both places in both cases. I am realising an additional wrinkle here, as I write, but can I say that, as an end-user, I would be perfectly happy if it is the OEM specified SMBIOS value which is copied into the new, blank Data Hub in this case. (In the unusual - almost unheard of?! - situation that the OEM SMBIOS and OEM Data Hub values are different and that you want to preserve this, then fine, switch off Automatic and specify all values manually!)
- Lastly, as it is currently, UpdateDataHub literally cannot be used in combination with OEM specified values: if you want to actually update the DataHub, i.e. using ProtocolOverrides/DataHub = true, then you _have_ to turn off Automatic and manually expand out, expressing everything twice, once in each section (therefore also having to manually express various stuff which would be worked out automatically in Automatic mode). If not, if you use UpdateDataHub with Automatic mode, you necessarily get missing values for system id and serial in the Data Hub, with no other way to fix it. (Which, unsurprisingly - but I have tested - matters and messes up iCloud accounts.) (NB Everything else in the Data Hub, except for just these two ids, already gets populated exactly as you would want when using the simpler, Automatic, OEM specified settings. This is important: _everything else_ already works as you would want for cloning system an existing system's ids, with just the simpler settings!) The upshot of all this (why I think this matters for OC users other than just me) is that you simply cannot express, using an OC config, the meaning "use _this_ system's existing ids, whatever they are, in SMBIOS and Data Hub"; instead, you have to _explicitly_ configure, twice (or three times, if including boot NVRAM) per id, the ids that you want to use, which obviously means they then have to be collected per machine and placed into the config. And this makes configuring OC to run e.g. Big Sur on legacy Macs whilst [not changing the identity of the user's machine](https://github.com/dortania/Opencore-Legacy-Patcher/issues/51) significantly enough to matter more complicated to configure than it otherwise would be!

**Comment by vit9696** [comment]
@Bmju, let me try to clarify things in short. Some of your points are right, but I believe you made a mistake in the others, which makes you confuse other things.

- SystemUUID in Generic section is essentially setting 3 values: SMBIOS, DataHub, NVRAM. For each the decision is taken individually, therefore `Use the OEM specified value if there is one` is correct in terms of not using any value for DataHub when there is none.
- Different SystemUUIDs in DataHub and SMBIOS do exist. They differ in byte order I believe, and this at least is true for hackintosh. This is the reason `UseRawUuidEncoding` option appeared in 0.6.6, allowing the to match when set to `YES`.
- For non-Apple hardware DataHub never contains SystemUUID. Therefore setting SystemUUID is a requirement for macOS functioning. It also means that there is no need to drop DataHub on Apple-hardware. Non-Apple OEM entries will work just fine.
- For Apple hardware DataHub does contain SystemUUID, therefore setting SystemUUID is not a requirement, but you cannot override it without reinstalling the whole protocol. It is the same for all other DataHub values. Therefore for Apple hardware it may mean that in certain cases specifying a non-empty SystemUUID (and other fields) is a requirement.
- There is no way to override `some` DataHub keys. Neither in Automatic, nor in non-Automatic modes. At all. The keys must all be dropped before they can be set.
- Attempting to read the UUID or Serial from SMBIOS and trying to apply it on DataHub (and NVRAM) is probably what you want me to do. However, I am not positive this is what we should do. It is extra code, which benefit is very slim and rare.

**Comment by Bmju** [comment]
@vit9696 

Thank you. I think I have, now (but only just, only recently) understood all that already! Your are right that I only understood your first point last, and as I was writing the above. (It is still perhaps not at all clear from the docs, or from an end user point of view, that this is what is happening!)

You are also right that (something like) the last one is what I am asking for. Because OC is _so_ close to doing exactly what is necessary to clone an existing Macs identity into all sections without having to explicitly specify the old id.

And you are right that I am asking for this for a different audience than the traditional OC audience; i.e. not Hackintosh, but using OC to run an unsupported OS on legacy Apple hardware.

Thinking about how this interacts with the existing logic (which I now, finally, understand - I think) there are some different possible suggestions:

----

?Add a note perhaps at the top of "10.2 Generic Properties":

> Note: For those settings which have a failsafe of _OEM specified or not installed_, the decision as to whether or not there is an existing OEM specified value applies separately to NVRAM, SMBIOS and Data Hub data. The existence of an OEM value in one does not necessarily entail the existence of an OEM value in another, nor will they be copied between each other where a value does not exist in all places.

Something like that?

---

?Then, either:

Add a flag to use SMBIOS values (where applicable; which is system-id, SystemSerialNumber and Model, I think?) to populate the Data Hub (i.e. therefore not any longer obeying the above warning, at least on a flag; the reason being, to work round the fact that existing DH values are discarded). The only motivation for this I agree is specifically for the use case of cloning id on a legacy Mac.

Or add (a) flag(s) to make it possible that the legacy values are NOT discarded when creating the new driver, but are copied in. (I assuming this is probably not even optionally available now because both protocols cannot be available at the same time? Or, generally, for a good reason which makes this not an easy option.)

**Comment by Bmju** [comment]
PS I do agree, having understood the above, it works to spec now, and any change is an upgrade, for one specific (but perhaps now more common) use-case. And that even for that use-case, there's another way to do it - though only by manually specifying values - anyway.

**Comment by Bmju** [comment]
Also (sorry) in a case where you _want_ to clone the machine id, it's arguable there's not much point writing to Data Hub anyway, since it's already right in the values that matter (system-id and SystemSerialNumber). The Data Hub will be then wrong, and inconsistent with the rest of the system, on `Model` (because that is present in the default Data Hub and won't change), but that's essentially how all of the systems using OC on legacy hardware have all been, so far, i.e. although the inconsistency is there, it doesn't seem to matter at all, to anything that matters.

**Comment by vit9696** [comment]
Will reopen until the decision is taken. Want to hear other's arguments.

CC @Andrey1970AppleLife @dhinakg  @mhaeuser @khronokernel @vandroiy2013

**Comment by Bmju** [comment]
(Thanks for the pings to others, and agreement to take more input! You did not actually re-opentho', at least as far as my GitHub is currently showing. Suspecting an oversight given your comment...?! If I'm not getting confused about something else....!!)

**Comment by Bmju** [comment]
I noticed another issue which potentially speaks in favour of having an option to clone the existing values when creating a Data Hub driver. Namely that the bottom two bytes of FSBFrequency vary from run to run (e.g. as shown in the screenshots in my OP of this issue, actually). I understand (I think I understand!) that this is because the CPU frequency has to be measured, and that a lot of effort (#448) has gone into getting this right.

I hope I'm not being to stupid to think that on a legacy machine which already has pre-existing, Apple-configured timing values, it's reasonable to want to prefer to keep these, even if/when one wants to fake other settings in the Data Hub.

I think I'm right in understanding that a different value means that one's system clock (normal, time of day clock) would run at a slightly different speeds? If so, this (v. minor) FSBFrequency issue arguably speaks against over-writing existing Data Hub values on legacy machines - at least without taking care to preserve this value?

I appreciate that just cloning the Data Hub would not be enough on its own to change the existing behaviour here anyway. (I think it would need yet another flag, or some other logic to not overwrite these values when existent and appropriate.)

Also, that all of this thread (including the above suggestions) is asking for gold-plating to achieve, via the Generic section and Automatic detection of OEM values, **things that can already be achieved anyway via the non-Automatic sections and manual config of measured values**. Notwithstanding that, I can't quite resist mentioning a new way I thought of, of phrasing the over-arching issue about the Data Hub here; which is that allowing cloning it would allow removing the asymmetry between Data Hub and SMBIOS/NVRAM which made the original (non-) issue look like a real issue in the first place! 😉 

I do totally agree that everything I want to do, for my own config, and everything I'm trying (perhaps unsuccessfully) to suggest that others might want to do, definitely all can be done already with the existing system, once all the relevant issues are understood.

**Comment by vit9696** [comment]
Currently the design behind PlatformInfo is not very consistent to the very least. Below I tries to express the key points behind the issues.

There are three areas that we control through PlatformInfo:
- SMBIOS
- DataHub
- NVRAM

The areas can be updated in automatic and manual modes. Automatic mode can be viewed as a variant of manual mode with certain fields pre-filled with well-known data based on SystemProductName, and is therefore not part of the main issue. For the time being let's ignore Automatic.

**DataHub** is not a key value structure but instead works like a tape storage:
- Iteration is from first to last, i.e. essentially it is a manual lookup over the so-called log storage.
- Writing any new entry simply appends the entry to the end of the log.
- It is possible to erase DataHub contents by reinstalling its protocol.

**SMBIOS** is also not a key-value structure, but it is closer. It is a serialised list of tables with mandatory keys and values.
- Iteration is from first to last but through the list of packed tables with field being an offset in the table.
- Modification is only possible through generating an entirely new blob with all relevant tables.

**NVRAM** is a key-value structure.
- Iteration is unnecessary thanks to random-access Get/Set methods.
- Modification is possible through Set method, but may fail in case of storage mismatch (RAM vs NV FLASH).

Based on the above, here is how the thing works currently:
- DataHub gets new values appended to the existing ones with macOS kernel using first found. Essentially this means that specified values do not replace existing ones.
- SMBIOS is regenerated from the ground, preserving known values and dropping all unknown tables, with the version forcibly set to 3.2. That said, specified values for SMBIOS are overrides, which replace existing ones.
- NVRAM has the values overwritten, unless they have different attributes. I.e. while normally specified values for NVRAM override the the existing ones, if the existing ones have NV FLASH type instead of RAM, they will not be overwritten.

Additional things to keep in mind:
- Changing SMBIOS version may affect the preferred UUID decoding order for some tools like dmidecode.
- All the three are treated independently, therefore if SMBIOS has the UUID but NVRAM and DataHub do not, then it is not propagated from SMBIOS to DataHub/NVRAM with Automatic=YES.
- The firmware may have existing SMBIOS, DataHub, and NVRAM entries that can be of use: InitialTSC, FSBFrequency, ARTFrequency, StartupPowerEvents, SMC information, memory status, Apple CPU type.


The above raises the following questions in my mind (feel free to add yours):

1. Should we provide an ability to override existing DataHub values? This gives the benefit of being able to change board information preserving current identifiers (e.g. UUIDs) on Mac hardware without knowing them. The downside is that some Mac users may be used to these identifiers not changing regardless of the values they specify in config.plist.
2. Should we forcibly delete NV FLASH variables from NVRAM and do our best to ensure that our variables do override all NVRAM variables? Currently failing to do this causes a RELEASE-visible warning, which makes me think we faced no machines where this is true yet.
3. Should we propagate select OEM values from SMBIOS (e.g. UUID) to DataHub and NVRAM with Automatic=YES? This can be done both implicitly (empty value) or explicitly (special value). The benefit of doing this is allowing to preserve fields that may be used for software activation. The downside is that some platforms have non-unique identifiers. They can be blacklisted, but I do not believe we have a list.
4. Should we use CPU frequency information from DataHub if it exists (InitialTSC, FSBFrequency, ARTFrequency)? I do not believe it can be wrong, but I am not fully positive we can always reliably calculate TSCFrequency from FSBFrequency. Likely we do though. There is a separate issue for that: https://github.com/acidanthera/bugtracker/issues/1484
5. Should we use OEM data from DataHub and SMBIOS in case DataHub is reinstalled and/or DataHub/SMBIOS are overridden in Automatic mode? The biggest questions are DataHub StartupPowerEvents, SMBIOS SMC info, SMBIOS Apple CPU type, SMBIOS RAM failure information. While currently all these are replaced with «fake» information, no obvious downside has ever been discovered. I may have missed some values.

My opinion on these questions is follows:

1. There is no easy way to do it, and I do not believe we should try. Our DataHub implementation may get key-value access, and we can surely preserve the existing DataHub information when reinstalling the protocol, but this raises more questions than answers, and the resulting protocol will not really work like DataHub.
2. Not until it actually is discovered. And if it is, decide case by case. Deleting NV variables may be unsafe. I do believe we should improve the docs better describing the behaviour. For all the three.
3. With a blacklist and explicit values it can be considered. We can specify a reserved `OEM` keyword for MLB, ROM, SystemSerialNumber, SystemUUID, and try to extract these values.
4. DataHub (before the reinstallation) should become just another CPU frequency provider, similarly to all the others we have. We can read this value and use without recalculation.
5. I do not believe there is a good benefit in doing this. Perhaps it can be decided in the future on case-by-case basis.

**Comment by Bmju** [comment]
Many thanks for very detailed explanation and info.

If I understand correctly the suggestion implied in the above (barring further comments) would be to implement 4 (probably) and 3 (possibly)? That sounds sufficient to address both my issues as they stand, and more generally a) both would be very useful for those who need (or want) to fiddle with some Data Hub settings on Apple hardware while preserving others, and b) using Data Hub info to get timing-free OEM specified CPU frequency where available also sounds ideal!

**Comment by Bmju** [comment]
May I add a couple more comments?

About whether to clone pre-existing DataHub values (maybe on a new flag) (~option 1) or whether to provide a new OEM flag-value for some fields, in order to propagate OEM SMBIOS values more widely, including into new, empty Data Hub (option 3). Given your new info, I wonder if there is any mileage in cloning any pre-existing OEM Data Hub values from the old protocol into the new protocol *after* any OC values have been set? That possibly might be a way to still do 1 (despite the quirks of the protocol) and have no need for 3? I have no objection to 3 if this still doesn't work! Am I'm not here considering whether the suggested new 'OEM' flag-value idea might provide useful features for NVRAM (although NVRAM is both less fiddly to work with and - apparently - much less important to successful software activation; so at least for that purpose any such new features might be less needed than they are for Data Hub?)

Re 3 and blacklisting known, non-unique hardware identifiers: might it be sufficient to say in the docs that the ability to preserve existing identifiers is intended for OC on legacy Mac hardware, and that the recommended (and only guaranteed safe) approach for non-Mac hardware is to use macserial + genuuid?

**Comment by Bmju** [comment]
Secondly, re option 3 and blacklisting known, non-unique hardware identifiers: might it be sufficient (or, at least, additionally helpful) to say in the docs that the ability to preserve existing identifiers is intended for OC on legacy Mac hardware, and that the recommended (and only guaranteed safe) approach for non-Mac hardware is to use `macserial` + `uuidgen` to generate all-new identifiers? (That said, I believe there's a debate around what IS safe to change on legacy Macs with the existing serials preserved: is it okay to fake board-id and board-product-name using Automatic=true, or is it somehow better/safer only to fake board-id (which currently I believe would require Automatic=false). Or even: is it not safe at all to use existing serials on legacy Macs with _any_ spoofing, even it's just board-id &/or name. (I could list different projects, claiming all three different positions!) That is all probably another issue, although it is in the same ballpark of saying what IS okay on legacy Macs, that might not be okay on non-Apple hardware.)

**Comment by mhaeuser** [comment]
1. Not sure whether it is needed, but it makes sense to me to preserve the original DataHub pointer and treat the protocol as a blob, just as done with SMBIOS. The original blob(protocol) would be carried from overriding in the platform update stage, where, after OC is done populating all its entries (which then take precedence), the original values are appended. The main issue I see with it is that we always need to override for consistent behaviour this way.
2. Agreed.
3. I'm still not sure I understood the blacklisted values, but I guess a warning could be sufficient. Ideally the user knows what he asks for when enabling OEM propagation.
4. Agreed.
5. Sorry, no idea.

**Comment by Bmju** [comment]
🎉 Brilliant - thank you for the updates!

**Comment by vit9696** [comment]
Please confirm it works correctly.

**Comment by Bmju** [comment]
Absolutely - have only just now had a chance to compile and test.

**DataHub** - works as expected and as wished, thank you! (Also as expected, to work as wanted in Data Hub requires `UseRawUuidEncoding=true`)
**NVRAM** - suspect minor bug introduced?
 - In 0.6.7 nvram `HW_MLB` is not being set even for _explicit_ user value at `PlatformInfo/Generic/MLB`; I would expect the behaviour for this to be unchanged, and in 0.6.6 it was set in this case. (Note: nvram `MLB` is still working as before and as expected.)
 - Additionally, I would have *expected* that using the new `OEM` string value for MLB would also set nvram `HW_MLB` (which it does not) as well as nvram `MLB` (which it does)?
 - Iff my expectation above is correct, then I think it would be natural to additionally allow `OEM` string value in `PlatformInfo/Generic/ROM` (because explicit value currently propagates to nvram `ROM` and `HW_ROM`, but empty value only propagates to nvram `ROM`, therefore we need a way to propagate detected OEM value to where explicit value goes)

**Comment by vit9696** [comment]
I believe I fixed the situation with `HW_MLB` in master when `OEM` is set in https://github.com/acidanthera/OpenCorePkg/commit/d8946d98f1af2517717ebdd14cc154c812c6256b.

I also introduced OEM ROM support for NVRAM in https://github.com/acidanthera/OpenCorePkg/commit/c97baf360beb66b2be8c653c0fa4241c73d9a211.

Please confirm works as designed.

**Comment by Bmju** [comment]
Checking - seeing some values I believe are wrong - checking again with new commits as of now

**Comment by Bmju** [comment]
<table>
<tr><th>NVRAM var</th><th>Explicit user value 0.6.7</th><th>Explicit user value 0.6.6</th><th>New `OEM` value 0.6.7</th><th>Empty string value 0.6.7</th><th>Empty string value 0.6.6</th></tr>
<tr><td>SSN</td><td>OEM - <i>possible new issue?</i> (expect user value)</td><td>OEM - <i>possible new issue?</i> (expect user value)</td><td>OEM</td><td>OEM</td><td>OEM</td></tr>
<tr><td>system-id</td><td>User</td><td>User</td><td>OEM</td><td>OEM</td><td>OEM</td></tr>
<tr><td>ROM</td><td>User</td><td>User</td><td>Literally "OEM" - <b>error</b> (borked value)</td><td>OEM</td><td>OEM</td></tr>
<tr><td>HW_ROM</td><td>User</td><td>User</td><td>Literally "OEM" - <b>error</b> (presence correct, but borked value)</td><td>Not present</td><td>Not present</td></tr>
<tr><td>MLB</td><td>OEM - <b>error (expect user value)</b></td><td>User</td><td>OEM</td><td>OEM</td><td>OEM</td></tr>
<tr><td>HW_MLB</td><td>Not present - <b>error</b> (expect no change from behaviour of explicit user value in 0.6.6)</td><td>User</td><td>Not present - <b>error</b> (expect presence or absence to behave like explicit user value)</td><td>Not present</td><td>Not present</td></tr>
</table>

**Comment by vit9696** [comment]
May I see the log with where you think you have errors? Make sure to use 5ae3b6f.

**Comment by vit9696** [comment]
Also include the config.plist you used in each case.

**Comment by Bmju** [comment]
I am not using the log files, therefore I 100% appreciate that as you need confirmatory information this is a completely non-standard bug report (or bug claim). I can only apologise. I am not sure how to provide the information you need in another way.

I am using my own tool which I was working on anyway to directly view the NVRAM vars in section `4D1EDE05-38C7-4A6A-9CC6-4BCCA8B38C14`, viewed within (or outside of) OC but prior to OS boot. (As I briefly mentioned when I wrote that entire [plist config tool](https://github.com/acidanthera/OpenCorePkg/pull/173) as a possible offer for OC, I was and am still planning to make this tool of mine more configurable via plist config - it _can_ currently load an OC-style config file `BootHelper.plist` (although I've compiled this version not to try), but after writing the tool to generate OC config .c and .h files, I have not yet come back to actually working on the customisation for this - hence the values to display currently need to be [configured in code](https://github.com/MikeBeaton/BootHelperPkg/blob/fb8e429e17e1b3dcb982e7ed575edde27dbd178b/Application/BootHelper/BootHelper.c#L161-L166) and recompiled, which I have done here.) The link to a version of the tool which displays the info summarised above is [here](https://github.com/MikeBeaton/BootHelperPkg/releases/download/0.2.9-oc1/bootx64.efi). The corresponding code is in [the `wip` branch](https://github.com/MikeBeaton/BootHelperPkg/tree/wip). I have just built, pushed and beta-released that version specifically for the purpose of this bug report. More or less obviously, I am sure, the file can be renamed to BootHelper.efi and installed as an OC tool (it works like that), or it can be placed as is into an /EFI/BOOT dir somewhere so that it can be run as a boot option inside or outside OC (it works like that). (This all works, I have been using it just to toggle NVRAM values and confirm OC version quickly for my own convenience for ages!) More importantly, for now, it is how I was actually viewing and reporting the above information. Since the output is coherent and sensible, I believe it is working. Example screenshot below from the tool run within OC 0.6.6, with config (from the below config file) using specifically identifiable ('OCSER', 'OCMLB', etc.) user specified values (unchanged value redacted):

![d6d9167e-4ef0-413f-b201-1e8402bfd43d](https://user-images.githubusercontent.com/25643187/107886523-39775a00-6ef8-11eb-81ae-ca9b1acfe227.jpeg)

The values shown, from the NVRAM section mentioned, are changing as I reported between versions and settings.

This is the relevant section of the OC config file which I was using, commenting keys in and out of it as appropriate, and copying over 0C 0.6.6 DEBUG (your build) files or OC 0.6.7 DEBUG (my build) files also as appropriate:

```
	<key>PlatformInfo</key>
	<dict>
		<key>Automatic</key>
		<true/>
		<key>CustomMemory</key>
		<false/>
		<key>Generic</key>
		<dict>
			<key>AdviseWindows</key>
			<false/>
			<key>MaxBIOSVersion</key>
			<true/>

			<key>#MLB</key>
			<string>{MY-OEM-VALUE-REDACTED}</string>
			<key>MLB</key>
			<string>C02305OCMLBF26L1N</string><!--...OCMLB...-->
			<key>#MLB</key>
			<string>OEM</string>
			<key>#MLB</key>
			<string></string>

			<key>ProcessorType</key>
			<integer>0</integer>

			<key>#ROM</key>
			<data>{MY-OEM-VALUE-REDACTED}</data><!--{MY-OEM-VALUE-REDACTED}-->
			<key>ROM</key>
			<data>EN6twN7e</data><!--10DEADC0DEDE-->
			<key>#ROM</key>
			<data>OEM</data>
			<key>#ROM</key>
			<data></data>

			<key>SpoofVendor</key>
			<false/>
			<key>SystemMemoryStatus</key>
			<string>Auto</string>
			<key>SystemProductName</key>
			<string>MacBookPro11,1</string>

			<key>#SystemSerialNumber</key>
			<string>{MY-OEM-VALUE-REDACTED}</string>
			<key>SystemSerialNumber</key>
			<string>C0OCSERNUM55</string><!--...OCSERNUM...-->
			<key>#SystemSerialNumber</key>
			<string>OEM</string>
			<key>#SystemSerialNumber</key>
			<string></string>

			<key>#SystemUUID</key>
			<string>{MY-OEM-VALUE-REDACTED}</string>
			<key>SystemUUID</key>
			<string>11223344-5566-7788-99AA-BBCCDDEEFF00</string>
			<key>#SystemUUID</key>
			<string>OEM</string>
			<key>#SystemUUID</key>
			<string></string>

		</dict>
		<key>UpdateDataHub</key>
		<true/>
		<key>UpdateNVRAM</key>
		<true/>
		<key>UpdateSMBIOS</key>
		<true/>
		<key>UpdateSMBIOSMode</key>
		<string>Create</string>
		<key>UseRawUuidEncoding</key>
		<true/>
	</dict>
```

Apologies - I know this probably looks like a crazy bug report. I don't think it actually is!

**Comment by vit9696** [comment]
Hmmm. Basically I need logs not because I need to confirm the bug, but rather I want to inspect what actually happens inside OC on your system. I do not have Apple hardware with OpenCore, so I cannot really reproduce what you have locally. 

Let's try to settle on some things first:

- OpenCore is unmodified and is a freshly compiled DEBUG version from https://github.com/acidanthera/OpenCorePkg/commit/5ae3b6f26768c9e498c1b038b57f9f6f8b42b404. For clarity, please use https://github.com/dortania/build-repo/releases/download/OpenCorePkg-5ae3b6f/OpenCore-0.6.7-DEBUG.zip, not self-compiled version.
- Whatever way you generate the config.plist, no duplicating keys exist, as they **are not allowed** in plist dicts. In fact, I will not promise OC actually parses such plists in a way you expect.
- What is SSN? OpenCore does not write anything to SSN. Neither 0.6.6, nor 0.6.7, not even any other version. Whatever this thing is, it is not OpenCore-made.

I am investigating the situation with MLB, as I was able to reproduce some problem with it on another machine, yet for now please clear these things.

**Comment by vit9696** [comment]
Master has MLB issue fixed, please provide an update with regards to previous comments.

**Comment by vit9696** [comment]
Regarding `Literally "OEM" - error (borked value)`:

You will need to explain how did you specify OEM in the `ROM` value. It should be string → OEM. In the following log line you should see 1 near OEM ROM:
```
OC: PlatformInfo auto %d OEM SN %d OEM UUID %d OEM MLB %d OEM ROM %d - %r
```

**Comment by Bmju** [comment]
ROM was my error, apologies. `<data>T0VNAA==</data>` or `<string>OEM</string>` both work. (Entirely my fault, but since `ROM` is - brand-newly - the only `OEM`-able value which is `<data>` not `<string>` by default, might be worth putting "`OEM` (or `T0VNAA==` in base 64)" at that point in the docs for `ROM` to make it easy for others to avoid this mistake?)

Checking others.

**Comment by vit9696** [comment]
There is little sense to further expand docs in regards to ROM, as it explicitly says that ROM is now `plist multidata`, i.e. can be specified with multiple types in the plist files depending on purpose.

**Comment by Bmju** [comment]
Latest commit https://github.com/acidanthera/OpenCorePkg/commit/55eb273cb111e7b4ec924e7de047bdec19d4b168 fixes all three MLB errors in my grid. Nice!

So with that and specifying `OEM` for `ROM` correctly, we now have:

<table>
<tr><th>NVRAM var</th><th>Explicit user value 0.6.7</th><th>Explicit user value 0.6.6</th><th>New `OEM` value 0.6.7</th><th>Empty string value 0.6.7</th><th>Empty string value 0.6.6</th></tr>
<tr><td>SSN</td><td>OEM - <i>possible new issue?</i> (expect user value)</td><td>OEM - <i>possible new issue?</i> (expect user value)</td><td>OEM</td><td>OEM</td><td>OEM</td></tr>
<tr><td>system-id</td><td>User</td><td>User</td><td>OEM</td><td>OEM</td><td>OEM</td></tr>
<tr><td>ROM</td><td>User</td><td>User</td><td>OEM</td><td>OEM</td><td>OEM</td></tr>
<tr><td>HW_ROM</td><td>User</td><td>User</td><td>OEM</td><td>Not present</td><td>Not present</td></tr>
<tr><td>MLB</td><td>User</b></td><td>User</td><td>OEM</td><td>OEM</td><td>OEM</td></tr>
<tr><td>HW_MLB</td><td>User</td><td>User</td><td>OEM</td><td>Not present</td><td>Not present</td></tr>
</table>

i.e. It's all fixed! 🎉

4D1EDE05-38C7-4A6A-9CC6-4BCCA8B38C14:SSN exists on my Mac (Late 2012 13" retina display MacBookPro10,12) by default, after an NVRAM reset, with no OC involved (image) and contains my normal, user visible, system serial number: the one that `SystemSerialNumber` corresponds to. So maybe it should be being set by OC in general? (As an addition, obviously no longer any kind of bug in this new code.)

![no-oc](https://user-images.githubusercontent.com/25643187/107890967-c845a000-6f13-11eb-9b40-f3cd1f0ef86a.jpeg)

I think my config fragment must be valid!? Since OC plist files are illegal unless every key is present, and since OC (recently) provides a way to comment out keys with `<key>#KeyName</key>`, I think logically the value of a commented out key plus it's value must be completely ignored (which is what I was relying on, and which btw works perfectly as I would have expected) or else there is no way to use commented out keys legally...?! FWIW - `ocvalidate` is perfectly happy with the above. And - of course - complains as I would expect if all versions of a key are commented out with #. (I have just checked, and `ocvalidate` does not complain about duplicate keys though... which I guess it ideally should. I know one should not overly rely on it - and it is - I agree - already extremely useful!) (Which reminds me, `ocvalidate` does not yet know that `OEM` is now a valid value for `SystemUUID`.) (Possibly cc @PMheart on these?)

**Comment by Bmju** [comment]
By the way, I looked through the entire list of my NVRAM using my tool in order to see what values were potentially relevant here - that's how I got my list of what to include! SSN is the _only_ other one in there that corresponds to any of the serials and ids we are discussing, except for `BID` and `HW_BID` which I omitted in the above as there is no change there.

**Comment by vit9696** [comment]
Hrm. Regarding `SSN`, create a separate issue I guess. Mention whether `HW_SSN` exists as well.

As for duplicating IDs, you cannot enter them with Xcode or any other editor. It will error. Basically the reason you were able to add them is due to editing the file manually as a text file. Since such a plist file is not a valid plist file, adding checks for this on ocvalidate level is not really desired. In fact, it is also difficult due to how OcXmlLib works.

**Comment by Bmju** [comment]
Okay - but I am editing it as a text file, and the commenting out is supported, by OC, to that extent? So, that that extent, it's valid, or at least, I am only expecting OC to behave per spec if I expect it to read the single uncommented out value, if there is one (or to act is if the value is missing, which it should not be, if there is none, but that would still be per spec, FWIW, although my config file then wouldn't be; or, certainly to behave in any way it wants if there is more than one)?

**Comment by Bmju** [comment]
Will make new issue re SSN/HW_SSN - tomorrow! 🤣 Thank you.
